creation/create_cvmfsexec_distros.sh

```diff
@@ -17,86 +17,113 @@ CVMFS_SOURCES=osg:egi:default
+    cvmfsexec_tarballs=$work_dir/cvmfsexec/tarballs
+    if [[ -f "$cvmfsexec_tarballs/.cvmfsexec_version" ]]; then
+        curr_ver=$(cat "$cvmfsexec_tarballs"/.cvmfsexec_version)
+        echo "Current version found: $curr_ver"
+    fi
+else
+    # if the cvmfsexec directory does not exist
+    mkdir -p $work_dir/cvmfsexec
+    # create a directory named tarballs under cvmfsexec directory
+    cvmfsexec_tarballs=$(mkdir -p $work_dir/cvmfsexec/tarballs -v | awk -F" " '{print $4}' | sed 's/^.//; s/.$//' )
+    chmod 755 $cvmfsexec_tarballs
 fi
 
-if [[ ! -d $cvmfsexec_tarballs ]]; then
-    mkdir -p $cvmfsexec_tarballs
+# otherwise, .cvmfsexec_version file does not exist from a previous upgrade or it's a first-time factory upgrade
+# check if the temp directory for cvmfsexec from previous run exists
+if ls -d $work_dir/cvmfsexec/cvmfsexec.tmp > /dev/null 2>&1 ; then
+    # cvmfsexec temp directory exists; reuse this
+    cvmfsexec_temp=$(ls -d $work_dir/cvmfsexec/cvmfsexec.tmp)
+else
+    # cvmfsexec temp directory does not exist; so create one
+    cvmfsexec_temp=$(mkdir -p $work_dir/cvmfsexec/cvmfsexec.tmp -v | awk -F" " '{print $4}' | sed 's/^.//; s/.$//' )  # remove the single quotes at the start and end of the string
+    chmod 755 "$cvmfsexec_temp"
 fi
 
-declare -a cvmfs_sources
-cvmfs_sources=($(echo $CVMFS_SOURCES | tr ":" "\n"))
+cvmfsexec_latest="$cvmfsexec_temp"/latest
+CVMFSEXEC_REPO="https://www.github.com/cvmfs/cvmfsexec.git"
+git clone $CVMFSEXEC_REPO "$cvmfsexec_latest" &> /dev/null
+latest_ver=$("$cvmfsexec_latest"/cvmfsexec -v)
+if [[ "$curr_ver" == "$latest_ver" ]]; then
+    # if current version and latest version are the same
+    echo "Current version and latest version of cvmfsexec are identical!"
+    if [[ -f "$cvmfsexec_tarballs/.cvmfsexec_version" ]]; then
+        echo "Using (existing) cvmfsexec version `cat "$cvmfsexec_tarballs"/.cvmfsexec_version`"
+    fi
+    echo "Skipping the building of cvmfsexec distribution tarballs..."
+    rm -rf "$cvmfsexec_latest"
+    exit 0
+else
+    # if current version and latest version are different
+    if [[ -z "$curr_ver" ]]; then
+        # $curr_ver is empty; first time run of factory upgrade
+        # no version info stored in work-dir/cvmfsexec/tarballs
+        echo "Building cvmfsexec distributions..."
+    else
+        # $curr_ver is not empty; subsequent run of factory upgrade (and not the first time)
+        echo "Found newer version of cvmfsexec..."
+        echo "Rebuilding cvmfsexec distributions using the latest version..."
+    fi
 
-declare -a machine_types
-machine_types=($(echo $SUPPORTED_TYPES | tr ":" "\n"))
+    # build the distributions for cvmfsexec based on the source, os and platform combination
+    cvmfsexec_distros="$cvmfsexec_temp"/distros
+    if [[ ! -d "$cvmfsexec_distros" ]]; then
+        mkdir -p "$cvmfsexec_distros"
+    fi
+
+    if [[ ! -d "$cvmfsexec_tarballs" ]]; then
+        mkdir -p "$cvmfsexec_tarballs"
+    fi
```

Comment:
`mkdir -p` is OK if the directory is already there. Does nothing, no need to check.
I think you created already above the tarballs directory

Reply:
Correct, the tarballs directory will be created at the beginning. Removed lines 86:88 as suggested.